File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;
vector<int> adj(101, -1);
vector<int> vis(101, 0);

int solve(int n) {
    queue<int> q;

    q.push(n);
    vis[n] = 1;
    int cnt = 0;
    while(!q.empty()) {
        int cur = q.front(); q.pop();
        cnt++;
        if (vis[adj[cur]] == 1) continue;
        vis[adj[cur]] = 1;
        q.push(adj[cur]);
    }

    return cnt;
}

int solution(vector<int> cards) {

    int ans = 0;
    
    for(int i=0; i<cards.size(); i++) {
        adj[i+1] = cards[i];
    }
    
    for(int j=1; j<cards.size(); j++) {
        int score = solve(j);
        bool flag = false;
        for(int k=j+1; k<=cards.size(); k++){
            if (vis[k] == 1) continue;
            int tmp = solve(k);
            ans = max(ans, score * tmp);
            flag = true;
        }
        if (!flag) ans = max(ans, 0);
        
        fill(vis.begin(), vis.begin() + cards.size() + 1, 0);
    }

    
    cout << ans;
        
    return ans;
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {

    ArrayList<Integer> adj = new ArrayList<>(Collections.nCopies(101, -1));
    ArrayList<Integer> vis = new ArrayList<>(Collections.nCopies(101, 0));
    
    public int solution(int[] cards) {
        for(int i=0; i<cards.length; i++) {
            adj.set(i+1, cards[i]);
        }
        ArrayList<Integer> counts = new ArrayList<>();
        for(int i=1; i<=cards.length; i++) {
            int cnt = 0;
            int cur = i;
            while(vis.get(cur) != 1) {
                vis.set(cur, 1);
                cur = adj.get(cur);
                cnt++;
            }
            counts.add(cnt);
        }
        
        counts.sort(Comparator.reverseOrder());
        
        return counts.size() == 1 ? 0 : counts.get(0) * counts.get(1);
    }
}